From 2df9172a09776d16694e2e833a35f633cbc157ac Mon Sep 17 00:00:00 2001
From: David Bauer <mail@david-bauer.net>
Date: Wed, 4 Feb 2026 21:57:52 +0100
Subject: [PATCH 1/3] wifi: mac80211: add flag for driver-buffering of PS

Add a new flag for drivers which retain frames of power save stations in
said driver or firmware.

This is necessary to be able to account for these pending frames when
calculating AQL limits.

Signed-off-by: David Bauer <mail@david-bauer.net>
---
 include/net/mac80211.h | 5 +++++
 net/mac80211/debugfs.c | 1 +
 2 files changed, 6 insertions(+)

--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -3074,6 +3074,10 @@ struct ieee80211_txq {
  * @IEEE80211_HW_SUPPORTS_NDP_BLOCKACK: HW can transmit/receive S1G NDP
  *	BlockAck frames.
  *
+ * @IEEE80211_HW_STORES_PS_FRAMES: Hardware/driver stores frames for
+ *	power-saving stations instead of returning them to mac80211 for
+ *	buffering.
+ *
  * @NUM_IEEE80211_HW_FLAGS: number of hardware flags, used for sizing arrays
  */
 enum ieee80211_hw_flags {
@@ -3135,6 +3139,7 @@ enum ieee80211_hw_flags {
 	IEEE80211_HW_HANDLES_QUIET_CSA,
 	IEEE80211_HW_STRICT,
 	IEEE80211_HW_SUPPORTS_NDP_BLOCKACK,
+	IEEE80211_HW_STORES_PS_FRAMES,
 
 	/* keep last, obviously */
 	NUM_IEEE80211_HW_FLAGS
--- a/net/mac80211/debugfs.c
+++ b/net/mac80211/debugfs.c
@@ -466,6 +466,7 @@ static const char *hw_flag_names[] = {
 	FLAG(HANDLES_QUIET_CSA),
 	FLAG(STRICT),
 	FLAG(SUPPORTS_NDP_BLOCKACK),
+	FLAG(STORES_PS_FRAMES),
 #undef FLAG
 };
 
